storm-core/src/jvm/org/apache/storm/daemon/supervisor/SyncProcessEvent.java

```diff
@@ -0,0 +1,669 @@
+        Map<String, Integer> approvedWorkers = localState.getApprovedWorkers();
+        Set<String> approvedIds = new HashSet<>();
+        if (approvedWorkers != null) {
+            approvedIds.addAll(approvedWorkers.keySet());
+        }
+        for (Map.Entry<String, LSWorkerHeartbeat> entry : idToHeartbeat.entrySet()) {
+            String workerId = entry.getKey();
+            LSWorkerHeartbeat whb = entry.getValue();
+            State state;
+            if (whb == null) {
+                state = State.notStarted;
+            } else if (!approvedIds.contains(workerId) || !matchesAssignment(whb, assignedExecutors)) {
+                state = State.disallowed;
+            } else if (supervisorData.getDeadWorkers().contains(workerId)) {
+                LOG.info("Worker Process {}as died", workerId);
+                state = State.timedOut;
+            } else if (SupervisorUtils.isWorkerHbTimedOut(now, whb, conf)) {
+                state = State.timedOut;
+            } else {
+                state = State.valid;
+            }
+            LOG.debug("Worker:{} state:{} WorkerHeartbeat:{} at supervisor time-secs {}", workerId, state, whb, now);
+            workerIdHbstate.put(workerId, new StateHeartbeat(state, whb));
+        }
+        return workerIdHbstate;
+    }
+
+    protected boolean matchesAssignment(LSWorkerHeartbeat whb, Map<Integer, LocalAssignment> assignedExecutors) {
+        LocalAssignment localAssignment = assignedExecutors.get(whb.get_port());
+        if (localAssignment == null || !localAssignment.get_topology_id().equals(whb.get_topology_id())) {
+            return false;
+        }
+        List<ExecutorInfo> executorInfos = new ArrayList<>();
+        executorInfos.addAll(whb.get_executors());
+        // remove SYSTEM_EXECUTOR_ID
+        executorInfos.remove(new ExecutorInfo(-1, -1));
+        List<ExecutorInfo> localExecuorInfos = localAssignment.get_executors();
+
+        if (localExecuorInfos.size() != executorInfos.size())
+            return false;
+
+        for (ExecutorInfo executorInfo : localExecuorInfos){
+            if (!localExecuorInfos.contains(executorInfo))
+                return false;
+        }
+        return true;
+    }
+
+    /**
+     * launch a worker in local mode.
+     */
+    protected void launchWorker(SupervisorData supervisorData, String stormId, Long port, String workerId, WorkerResources resources) throws IOException {
+        // port this function after porting worker to java
+    }
+
+    protected String getWorkerClassPath(String stormJar, Map stormConf) {
+        List<String> topoClasspath = new ArrayList<>();
+        Object object = stormConf.get(Config.TOPOLOGY_CLASSPATH);
+
+        if (object instanceof List) {
+            topoClasspath.addAll((List<String>) object);
+        } else if (object instanceof String){
+            topoClasspath.add((String)object);
+        }else {
+            //ignore
```

Comment:
Can we throw an exception here if object is not null?  If TOPOLOGY_CLASSPATH is something else we have an error in our code somewhere.